src/components/countries/CountryForm.tsx

```diff
@@ -0,0 +1,149 @@
+import React from 'react'
+import * as yup from 'yup'
+import { useRouter } from 'next/router'
+import { FormikHelpers } from 'formik'
+import { useMutation } from 'react-query'
+import { Button, Grid } from '@mui/material'
+import { AxiosError, AxiosResponse } from 'axios'
+import { useTranslation } from 'next-i18next'
+import { makeStyles } from '@mui/styles'
+
+import { routes } from 'common/routes'
+import { AlertStore } from 'stores/AlertStore'
+import { ApiErrors, isAxiosError, matchValidator } from 'service/apiErrors'
+import { useCreateCountry, useEditCountry } from 'service/restRequests'
+import { CountryInput, CountryResponse } from 'gql/countries'
+import { name } from 'common/form/validation'
+import GenericForm from 'components/common/form/GenericForm'
+import SubmitButton from 'components/common/form/SubmitButton'
+import FormTextField from 'components/common/form/FormTextField'
+import LinkButton from 'components/common/LinkButton'
+
+const useStyles = makeStyles({
+  container: {
+    maxWidth: '700px',
+    margin: '80px auto 0 auto',
+  },
+})
+
+const validationSchema: yup.SchemaOf<CountryInput> = yup.object().defined().shape({
+  name: name.required(),
+  countryCode: name.required(),
+})
+
+const defaults: CountryInput = {
+  name: '',
+  countryCode: '',
+}
+
+type CountryFormProps = {
+  initialValues?: CountryInput
+  id?: string
+}
+
+export default function CountryForm({ initialValues = defaults, id }: CountryFormProps) {
+  //if (id) -> edit form, else -> create form
+  const { t } = useTranslation('countries')
+
+  const createMutation = useMutation<
+    AxiosResponse<CountryResponse>,
+    AxiosError<ApiErrors>,
+    CountryInput
+  >({
+    mutationFn: useCreateCountry(),
+    onError: () => AlertStore.show(t('alerts.new-row.error'), 'error'),
+    onSuccess: () => AlertStore.show(t('alerts.new-row.success'), 'success'),
+  })
+  type CountryEditInput = {
+    id: string
+    data: CountryInput
+  }
+  const editMutation = useMutation<
+    AxiosResponse<CountryResponse>,
+    AxiosError<ApiErrors>,
+    CountryEditInput
+  >({
+    mutationFn: useEditCountry(),
+    onError: () => AlertStore.show(t('alerts.edit-row.error'), 'error'),
+    onSuccess: () => AlertStore.show(t('alerts.edit-row.success'), 'success'),
+  })
+  const router = useRouter()
+  const classes = useStyles()
+
+  const onSubmit = async (
+    values: CountryInput,
+    { setFieldError, resetForm }: FormikHelpers<CountryInput>,
+  ) => {
+    try {
+      const data = {
+        name: values.name,
+        countryCode: values.countryCode.toLocaleUpperCase(),
```

Comment:
Good call to uppercase the country code :raised_hands:

Reply:
I probably should of done that on the back-end too.

Reply:
Good idea